src/upsert.rs

```diff
@@ -68,60 +72,79 @@ pub trait FindexUpsert<
     /// - `additions`   : values to indexed for a set of keywords
     /// - `deletions`   : values to remove from the indexes for a set of
     ///   keywords
+    ///
+    /// Returns a map of keywords to booleans indicating whether the keyword
+    /// was already present in the database.
     async fn upsert(
         &mut self,
         master_key: &KeyingMaterial<MASTER_KEY_LENGTH>,
         label: &Label,
         additions: HashMap<IndexedValue, HashSet<Keyword>>,
         deletions: HashMap<IndexedValue, HashSet<Keyword>>,
-    ) -> Result<(), Error<CustomError>> {
+    ) -> Result<HashMap<Keyword, bool>, Error<CustomError>> {
         check_parameter_constraints::<CHAIN_TABLE_WIDTH, BLOCK_LENGTH>();
 
         let mut rng = CsRng::from_entropy();
         let k_uid = master_key.derive_kmac_key::<KMAC_KEY_LENGTH>(ENTRY_TABLE_KEY_DERIVATION_INFO);
         let k_value = master_key.derive_dem_key(ENTRY_TABLE_KEY_DERIVATION_INFO);
 
-        let mut new_chains =
-            HashMap::<KeywordHash, HashMap<IndexedValue, BlockType>>::with_capacity(
-                additions.len() + deletions.len(),
-            );
+        // a reverse map of Keyword UIDs (e.g. hashes) to Keywords
+        let mut uid_to_keyword = HashMap::<Uid<UID_LENGTH>, Keyword>::new();
```

Comment:
You can reserve the adequate space here.
```suggestion
        let mut uid_to_keyword = HashMap::<Uid<UID_LENGTH>, Keyword>::with_capacity(additions.len() + deletions.len());
```